### ISSUE TITLE
Assertion failure in vm/excep.cpp when running CoreFX tests

### ISSUE THREAD

**Issue by kouvel** [dotnet/coreclr#1595]
Discovering: System.Diagnostics.Tracing.Telemetry.Tests
Discovered:  System.Diagnostics.Tracing.Telemetry.Tests
Starting:    System.Diagnostics.Tracing.Telemetry.Tests

Assert failure(PID 31664 [0x00007bb0], Thread: 31681 [0x7bc1]): personalityRoutine != NULL
    File: /home/kouvel/e/clr/coreclr/src/vm/excep.cpp Line: 8264
    Image: /home/kouvel/test/corefxtests/test/corerun

**** MessageBox invoked, title 'corerun - Assert Failure (PID 31664, Thread 31681/0x7bc1)' ****
  personalityRoutine != NULL

coreclr/src/vm/excep.cpp, Line: 8264

**Comment by gkhanna79** [comment]
@janvorli Could this be due to bad/incorrect unwindinfo?

**Comment by janvorli** [comment]
@gkhanna79 I am sorry for a late reply, I have missed your question before. The answer is that it is possible, but to be sure, I would need to take a look in the debugger - it should be very easy to see from the disassembly at pCodeInfo->GetStartAddress() and the pUnwindInfo->SizeOfProlog when the assert is hit.
@kouvel - is it easy to reproduce?

**Comment by kouvel** [comment]
Yea it was reproducing every time when running tests from that assembly.

**Comment by janvorli** [comment]
Great, I was able to repro it. I am looking into it.

**Comment by janvorli** [comment]
The manged function that is causing this issue here is the System.GC.KeepAlive. The function IL is as follows:
```IL
IL to import:
IL_0000  2a                ret
```

And so the code that jitter generates is just this single instruction:
```asm
retq
```

The unwind info generated by the jitter is:
```
Unwind Info:
  >> Start offset   : 0x000000 (not in unwind data)
  >>   End offset   : 0x000001 (not in unwind data)
  Version           : 1
  Flags             : 0x00
  SizeOfProlog      : 0x00
  CountOfUnwindCodes: 0
  FrameRegister     : none (0)
  FrameOffset       : N/A (no FrameRegister) (Value=0)
  UnwindCodes       :
```

**Comment by janvorli** [comment]
The RtlVirtualUnwind returns personality routine only when the instruction pointer is in the body of the function, that means neither in prologue nor in epilogue of the function. 
The purpose of the code that asserts is to check in debug build that the managed function has a personality routine. 
So it gets the starting address of the function, adds prologue size and assumes that that address is in the body of the function. But for our problematic function with single return, it is not true, since there is no body. There can be potentially other forms of functions with no body that would fail in this assert.
On Windows, we can get the same assert, but the IsIPInEpilog function is called very rarely - from Thread::IsSafeToInjectThreadAbort, so the probability that it hits the problematic case is very low.
But on Unix, the IsIPInEpilog is called very frequently - every time when we are suspending runtime for GC. 
I believe a correct fix is to remove the block with this invalid assert.

**Comment by janvorli** [comment]
Fixed by #1654
